File: python/exer07d_data_server.py
'''
THE DATA SERVER PROBLEM
Version D: Solving the problem using a blocking queue
'''

import time
import threading
from queue import Queue



def check_auth_user():
    print('[   Auth   ] Start')
    # Send request to authenticator, check permissions, encrypt, decrypt...
    time.sleep(20)
    print('[   Auth   ] Done')



def process_files(lst_file_name: list, blkq: Queue):
    for file_name in lst_file_name:
        # Read file
        print('[ ReadFile ] Start', file_name)
        time.sleep(10)
        print('[ ReadFile ] Done ', file_name)

        blkq.put(file_name) # You may put file data here

        # Write log into disk
        time.sleep(5)
        print('[ WriteLog ]')



def process_request():
    lst_file_name = [ 'foo.html', 'bar.json' ]
    blkq = Queue()

    # The server checks auth user while reading files, concurrently
    threading.Thread(target=process_files, args=(lst_file_name, blkq)).start()
    check_auth_user()

    # The server waits for completion of loading files
    for _ in range(len(lst_file_name)):
        blkq.get()

    print('\nNow user is authorized and files are loaded')
    print('Do other tasks...\n')



process_request()

File: csharp/exercise/exer07d-data-server.cs
/*
 * THE DATA SERVER PROBLEM
 * Version D: Solving the problem using a blocking queue
 */
using System;
using System.Collections.Concurrent;
using System.Threading;



class Exer07D : IRunnable
{
    public void run()
    {
        var server = new DataServer();
        server.processRequest();
    }


    private class DataServer
    {
        public void processRequest()
        {
            var lstFileName = new String[] { "foo.html", "bar.json" };
            var queue = new BlockingCollection<string>();

            // The server checks auth user while reading files, concurrently
            new Thread(() => processFiles(lstFileName, queue)).Start();
            checkAuthUser();

            // The server waits for completion of loading files
            for (int i = lstFileName.Length; i > 0; --i)
            {
                queue.Take();
            }

            Console.WriteLine("\nNow user is authorized and files are loaded");
            Console.WriteLine("Do other tasks...\n");
        }


        // This task consumes CPU (and network bandwidth, maybe)
        private void checkAuthUser()
        {
            Console.WriteLine("[   Auth   ] Start");
            // Send request to authenticator, check permissions, encrypt, decrypt...
            Thread.Sleep(20000);
            Console.WriteLine("[   Auth   ] Done");
        }


        // This task consumes disk
        private void processFiles(String[] lstFileName, BlockingCollection<string> queue)
        {
            foreach (var fileName in lstFileName)
            {
                // Read file
                Console.WriteLine("[ ReadFile ] Start " + fileName);
                Thread.Sleep(10000);
                Console.WriteLine("[ ReadFile ] Done  " + fileName);

                queue.Add(fileName); // You may put file data here

                // Write log into disk
                Thread.Sleep(5000);
                Console.WriteLine("[ WriteLog ]");
            }
        }
    }
}
